Good Bye, BOJ 2020!

mu-hun opened the pull request:
https://www.acmicpc.net/contest/view/578

mu-hun committed:
Good Bye, BOJ 2020!

mu-hun committed:
Check time limit

maximum time complexity is combinations(32, 3)

mu-hun committed:
WIP: 20530 is timeout

mu-hun committed:
Use dict

mu-hun committed:
Remove useless brace

mu-hun committed:
Put trees in dictionary

key is border node(root), values is children of root

x86chi commented:
BOJ에서 [런타임 오류(ValueError)가 나왔다.](http://boj.kr/06a417de008145acba5b55ea472e1c61)

[생각하지 않은 케이스에 ValueError 를 일으키도록](https://github.com/x86chi/problem-solving/blob/f47ce5b2ffddbffa7db25b875ad714b5712e1023/acmicpc.net/problem/20530.py#L58-L59) 설계했기 때문에, 입력 케이스 중에 반례가 나온 것 같다.

x86chi commented:
거의 다 푼 것 같은데.. BOJ 채점 서버에서 [약 20초 후에 틀렸다고](http://boj.kr/4bd78dbba097435497051dfd4f97ca42) 알려준다. [go 라는 내부 함수에서](https://github.com/x86chi/problem-solving/blob/66a0f8/acmicpc.net/problem/20530.py#L43-L61) 문제가 있는 것 같다.

경계 노드와 자식을 `branchs` 이름으로 매핑한 뒤 탐색을 하는 걸 생각해냈다. 매핑된 데이터의 값은 튜플로 정리했다.

> 예시 : `{2: (4, 5), 3: (6, 7)}`

---

> [Gravekeeper 님 디스코드에서 질문으로 남겼다.](https://discord.com/channels/155036236342362114/744508844343951432/797781450287218739)

x86chi commented:
> https://www.acmicpc.net/board/view/65861

해당 문제 페이지에 질문 남겼다.

x86chi commented:
## 어떻게 접근했었나?

<img width="416" alt="Screen Shot 2021-03-20 at 1 36 40 PM" src="https://user-images.githubusercontent.com/5278201/111860124-bb8bf000-8988-11eb-8da2-4e3da07eab39.png">

> https://youtu.be/UEGtFwGD1fU?t=759

1. 그래프에서 간선이 1인 정점을 계속 지워나가면, 마지막에 하나만 남게 됩니다. 
2. 지워지는 정점과 가장 가까운 사이클 위의 점으로 트리 부분을 구분할 수 있습니다." 가 내가 참조한 해설인데

두 조건에서 내가 오해한 것이, 간선이 1인 정점의 그 간선을 모두 지우면 모든 트리가 지워진다고 생각했다.

<img src="https://user-images.githubusercontent.com/5278201/111860193-4a007180-8989-11eb-8161-0f1e374a112c.jpg" width="200px">

이런 방식으로 푼 결과 위 사진처럼 사이클에서 2레벨 떨어진 서브 트리를 사이클로 판단하게 되었다.

## 올바른 접근법

여러 차례 간선이 1인 정점을 지워야 한다.

| 그래프 | 첫번째 간선이 1인 정점 지우기 | 두번째 지우기 |
| :--: | :--: | :--: |
| ![1](https://user-images.githubusercontent.com/5278201/111859744-62bb5800-8986-11eb-9434-49de1264eb50.jpg) | ![2](https://user-images.githubusercontent.com/5278201/111860004-09ecbf00-8988-11eb-80d0-5b0a2b0db05f.jpg) | ![3](https://user-images.githubusercontent.com/5278201/111860011-11ac6380-8988-11eb-86a8-31101bd5c157.jpg) |
| 지워진 정점 | 6, 7, 8, 9 | 4 |

mu-hun committed:
Solved 20530

이 문제 풀이는 백준 채점 서버에서 제한 시간을 초과합니다:
http://boj.kr/3b49c7c5a3044db69b0ca9b3680f7f07

x86chi commented on acmicpc.net/problem/20530.py:
```diff
@@ -21,69 +22,131 @@ def make_graph(lines: List[Path]):
+        "queries": ((2, 4), (4, 5), (1, 4), (6, 2)),
+        "result": (True, True, False, False)
+    },
+    {
+        "raw_paths": ((1, 2), (1, 3), (2, 4), (2, 5), (3, 6), (3, 7), (2, 3), (4, 8), (4, 9), (3, 6), (3, 7)),
+        "paths": {1: {2, 3}, 2: {1, 4, 5, 3}, 3: {1, 6, 7, 2}, 4: {2, 8, 9}, 5: {2}, 6: {3}, 7: {3}, 8: {4}, 9: {4}},
+        "cycle_path": {1: {2, 3}, 2: {1, 3}, 3: {1, 2}},
+        "tree_path": {2: {4, 5}, 3: {6, 7}, 4: {2, 8, 9}, 5: {2}, 6: {3}, 7: {3}, 8: {4}, 9: {4}},
+        "queries": ((2, 4), (2, 8), (9, 8), (5, 8)),
+        "result": (True, True, True, True)
+    },
```
테스트 케이스에 사이클에서 2 레벨 떨어진 리프 노드를 갖는 그래프를 추가했다. 문제의 엣지 케이스를 이해서 풀이를 고쳐봤지만 백준 채점 서버에서 제한 시간을 초과한다. 실행 시간을 줄일 방법을 찾아보자.

mu-hun committed:
Reduce excute time to n from 2n

mu-hun committed:
Check time limit

x86chi commented:
문제를 거의 풀어두어서 Draft PR 이라고 보기 어렵다. 테스트 케이스에 시간 제한 넣어두었으니 정식 PR 로 전환합니다.
